consul/kv_consul.go

```diff
@@ -424,12 +408,18 @@ func (kv *consulKV) CompareAndSet(
 		pair.ModifyIndex = 0
 	}
 
-	ok, _, err := kv.client.KV().CAS(pair, nil)
+	ok, _, err := kv.client.CompareAndSet(kvp.Key, kvp.Value, pair, nil)
 	if err != nil {
 		return nil, err
 	}
 
 	if !ok {
+		kvp, getErr := kv.Get(pair.Key)
```

Comment:
we already do a similar Get and byte compare in the consulClient CompareAndSet implementation. We can avoid one extra Get operation by returning the kvp value in its implementation.